### PR TITLE
Add GoogLeNet link

### PR DESCRIPTION
**Pull request by ronekko**
I added pretrained GoogLeNet link to chainer.links.model.vision .

### CODE REVIEW

**Comment by rezoo** [comment]
Thank you for your PR, and sorry for the late review.
Now I would like to review your PR.

File: chainer/links/model/vision/googlenet.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        else:
+            x = x[:, :, 16:240, 16:240]
+        # Set volatile option to ON to reduce memory consumption
+        x = Variable(self.xp.asarray(x), volatile=flag.ON)
+        y = self(x, layers=['prob'])['prob']
+        if oversample:
+            n = y.data.shape[0] // 10
+            y_shape = y.data.shape[1:]
+            y = reshape(y, (n, 10) + y_shape)
+            y = sum_function(y, axis=1) / 10
```
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: I suppose that you decided to use `sum_function` to avoid overriding.
How about using `from chainer.functions.math import sum` and replacing it with `sum.sum(y, axis=1)`?

File: docs/source/reference/links.rst [hunk 2]
```diff
@@ -191,6 +191,13 @@ where ``fc7`` denotes a layer before the last fully-connected layer.
 Unlike the usual links, these classes automatically load all the
 parameters from the pre-trained models during initialization.
 
+GoogLeNet
+~~~~~~~~~
+.. autoclass:: GoogLeNet
+   :members:
+
+.. autofunction:: chainer.links.model.vision.googlenet.prepare
+
```
Comment (rezoo) [inline docs/source/reference/links.rst]: I'm curious why you decided to place `GoogLeNet` at the top instead of `VGG16Layers`.
My feeling is that VGG model is the most commonly used, and GoogLeNet is the next.

If you think that these models should be arranged alphabetically, it seems better to rearrange the other models.

**Comment by rezoo** [comment]
I have one additional question: did you check whether the output is equal to the original one?

File: chainer/links/model/vision/googlenet.py [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+from chainer.links.connection.inception import Inception
+from chainer.links.connection.linear import Linear
+from chainer.serializers import npz
+from chainer.utils import imgproc
+from chainer.variable import Variable
+
+
+class GoogLeNet(link.Chain):
+
+    """A pre-trained GoogLeNet model provided by BVLC [1].
```
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: I appended a draft of the description about GoogLeNet.

File: chainer/links/model/vision/googlenet.py [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+            it automatically downloads the caffemodel from the internet.
+            Note that in this case the converted chainer model is stored
+            on ``$CHAINER_DATASET_ROOT/pfnet/chainer/models`` directory,
+            where ``$CHAINER_DATASET_ROOT`` is set as
+            ``$HOME/.chainer/dataset`` unless you specify another value
+            as a environment variable. The converted chainer model is
+            automatically used from the second time.
+            If the argument is specified as ``None``, all the parameters
+            are not initialized by the pre-trained model, but the default
+            initializer used in the original paper, i.e.,
```
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: I'm sorry that I'm not careful about the default initialization.
The provider of the caffe model says that the original GoogLeNet used Gaussian for initialization, however I cannot find such description in the original paper.
So I want to have the default initilizer according to the caffe model's one, however it is neither Gaussian nor Glorot.

In [train_val.prototxt](https://github.com/BVLC/caffe/blob/master/models/bvlc_googlenet/train_val.prototxt#L63) of the caffe model, "xavier" with default parameter is used for `weight_filler` .
When "xavier" is used with `variance_norm: AVERAGE` parameter, it means that [the weights are initialized by `Uniform(-s, s), s = sqrt(6 / (fan_in + fan_out))` ](https://github.com/BVLC/caffe/blob/master/include/caffe/filler.hpp#L154), which is the Glorot's initializer.
However, [the default parameter is `variance_norm: FAN_IN` ](https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto#L61) and then [the weights are initialized by `s = sqrt(3 / fan_in)` ](https://github.com/BVLC/caffe/blob/master/include/caffe/filler.hpp#L160).

File: chainer/links/model/vision/googlenet.py [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+            ``chainer.initializers.GlorotUniform(scale=1.0)``.
+
+    Attributes:
+        available_layers (list of str): The list of available layer names
+            used by ``__call__`` and ``extract`` methods.
+
+    """
+
+    image_mean = numpy.array([104, 117, 123], dtype=numpy.float32)  # BGR
```
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: For the first question, I sometimes need the mean values of RGB.
So it is useful if `image_mean` was explicitly defined when someone want to use it.
Although we provide `prepare` function, I think there is a case that we cannot use it.
For example in my case of fine-tuning, I make an HDF5 file of dataset and store images into it with some preprocessing, in order to avoid repetition of the same preprocessing every time the images are loaded.
And in this situation I manually have to subtract the mean RGB from partially preprocessed images.

For the second point I agree with you, but I am not sure which of `[104.0, 117.0, 123.0]` or `[103.939, 116.779, 123.68]` is appropriate. It seems that the former is correct, since in [train_val.prototxt](https://github.com/BVLC/caffe/blob/master/models/bvlc_googlenet/train_val.prototxt#L34-L36) the mean RGB is defined as `[104, 117, 123]` .
I checked the accuracy for both cases on ILSVRC 2012 validation set, however the difference was negligible.
Here is the top-1, top-5 correct rate without oversampling and top-1, top-5 with oversampling for each case.

- `[104.0, 117.0, 123.0]`
    - 0.67376, 0.88168, 0.68622 0.89128
- `[103.939, 116.779, 123.68]`
    - 0.67378, 0.88168, 0.68622 0.89116

File: chainer/links/model/vision/googlenet.py [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+
+    def __init__(self, pretrained_model='auto'):
+        if pretrained_model:
+            # As a sampling process is time-consuming,
+            # we employ a zero initializer for faster computation.
+            kwargs = {'initialW': constant.Zero()}
+        else:
+            # employ default initializers used in the original paper
+            kwargs = {'initialW': uniform.GlorotUniform(scale=1.0)}
```
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: I investigated closely the initialization of the caffemodel.
It seems that they actually used Uniform(-s, s) with `s = sqrt(3 / fan_in)` .
Since [ `variance_norm` of `weight_filler` is not specified](https://github.com/BVLC/caffe/blob/master/models/bvlc_googlenet/train_val.prototxt#L63) and [the default value of `VrianceNorm` is `FAN_IN` ](https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto#L61), then

```
    weight_filler {
      type: "xavier"
    }
```

turns out [s = sqrt(3 / fan_in)](https://github.com/BVLC/caffe/blob/master/include/caffe/filler.hpp#L160) and this is `LeCunUniform(scale=1.0)` in Chainer.

File: docs/source/reference/links.rst (hunk 2 above)
Comment (ronekko) [inline docs/source/reference/links.rst]: I moved GoogLeNet next to VGG.
(Since the original order of VGG and ResNet can be seen as from weak model to powerful one for me, so I put GoogLeNet first as the weakest model.)

**Comment by ronekko** [comment]
> I have one additional question: did you check whether the output is equal to the original one?

Yes I did if you mean comparison between outpus of chainer's CaffeFunction and this PR's. These outputs are exactly same (i.e. `np.array_equal()` is True).
On the other hand, I did not compare between the outputs of caffe's and this PR's, because I did not have caffe environment.
Alternatively, I tested the accuracy on ILSVRC 2012 validation set by the model of this PR.
The results of top-1 and top-5 correct rate for single center crop (i.e. `oversample=False` of `model.prepare()` ) are 67.4% and 88.2%, which is slightly worse than the [officially reported performance](https://github.com/BVLC/caffe/tree/master/models/bvlc_googlenet) of 68.7% and 88.9%.

**Comment by ronekko** [comment]
Thank you so much for your review.
I updated the code and answered your points.
Could you check them?

**Comment by ronekko** [comment]
I rebased this branch to the master in order to include the latest code.

File: chainer/links/model/vision/googlenet.py (hunk 5 above)
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: > For example in my case of fine-tuning, I make an HDF5 file of dataset and store images into it with some preprocessing, in order to avoid repetition of the same preprocessing every time the images are loaded.

In this case, to reduce the size and IO access, I think it would be better to store the image with uint8 into hdf5, and write the code as follows:
```python
from chainer.links.model.vision.googlenet import prepare

class Dataset(dataset_mixin):
    def __init__(self, path):
        self.db = h5py.File(path, 'r')
        self.dataset = self.db["image"]

    def get_example(self, i):
        return prepare(self.dataset[i])
```
Although I did not have the opportunity to use image_mean, but I think there is no problem to add `image_mean` to the class variable. However, if you really want to do this, it would be better to add `image_mean` to the other classes like VGG and ResNet, and explicitly describe these variables to the document.

Regarding the average, there is no problem if the original uses discrete values.

File: chainer/links/model/vision/googlenet.py (hunk 4 above)
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: I understood the situation. At least we cannot say that the same initializer is used in the caffemodel. How about replacing the sentence as follows?

If the argument is specified as ``None``, all the parameters are not initialized by the pre-trained model, but the default initializer denoted as ~

File: chainer/links/model/vision/googlenet.py (hunk 6 above)
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: Thanks for the investigation. How about using `LeCunUniform`, and describing it to the document?

File: chainer/links/model/vision/googlenet.py (hunk 1 above)
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: I see.

**Comment by rezoo** [comment]
I finished the review. In my opinion, if you want to add `image_mean` to googlenet, it would be better to add them to the other models in the same manner. (In that case it would be appropriate that we first merge a PR without `image_mean`, and make another PR that adds `image_mean` to all the pre-trained models).

File: chainer/links/model/vision/googlenet.py (hunk 4 above)
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: I fixed the sentence.

File: chainer/links/model/vision/googlenet.py (hunk 6 above)
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: I added docs and comments in order to explain why LeCunUniform is used here.

File: chainer/links/model/vision/googlenet.py (hunk 5 above)
Comment (ronekko) [inline chainer/links/model/vision/googlenet.py]: > In this case, to reduce the size and IO access, I think it would be better to store the image with uint8 into hdf5

Thank you for sharing your example code. I've been also doing the similar way (although I'm using `fuel` instead of chainer's dataset & iterator) and therefore I cannot subtract beferehand the mean pixel values, which are float value in general. In my observation, when the storage is HDD, IO bottleneck is problematic. However, when I use SSD, CPU bottleneck is dominant (especially jpg decoding and resizing).

Anyway, I do not use `image_mean` this time. As you pointed out, it should be added in another PR.

**Comment by ronekko** [comment]
Thank you. I reflected your points.

File: chainer/links/model/vision/googlenet.py (hunk 3 above)
Comment (rezoo) [inline chainer/links/model/vision/googlenet.py]: Good.

File: docs/source/reference/links.rst (hunk 2 above)
Comment (rezoo) [inline docs/source/reference/links.rst]: I see.
